### PR TITLE
efiperun: Add support for (constant) UEFI variables

### PR DESCRIPTION
**Pull request by ghost**
This patch extends efiperun with basic support for UEFI variables. With it applied a debug module can create a variable during initialization, like this:

``` C++
static void
init()
{
    static EFI_GUID VENDOR_GUID = { ... };

    static unsigned char oem_model_data[4] = {
        0xbc, 0x01, 0x00, 0x00
    };

    set_variable(&VENDOR_GUID,
                 (CHAR16 *)u"OemModelVar",
                 oem_model_data,
                 4,
                 EFI_VARIABLE_NON_VOLATILE
                 | EFI_VARIABLE_BOOTSERVICE_ACCESS
                 | EFI_VARIABLE_RUNTIME_ACCESS);
}
```

and the variable will be available to other modules through calls to Runtime Services' `GetVariable`.

This is useful (required, even) when debugging modules that expect to run in a specific context initialized by other modules.

### CODE REVIEW

File: efiperun/efihooks.cpp [hunk 1]
```diff
@@ -277,6 +285,32 @@ void install_protocol(EFI_GUID* guid,EFI_HANDLE handle,void* interface)
...
 
+void* get_variable(EFI_GUID* guid,char16_t* name,unsigned int* data_size,uint32_t *attributes=NULL)
+{
+	for (auto& elem : as_range(g_variables.equal_range(*guid)))
+	{
+		auto& el_name=elem.second.first;
+		auto& el_var_data=elem.second.second;
+		if (el_name->compare(name)==0)
+		{
+			*data_size=el_var_data->data_size;
```
Comment (jethrogb) [inline efiperun/efihooks.cpp]: Can you add a null pointer check here?

File: efiperun/main.h [hunk 2]
```diff
@@ -45,9 +45,12 @@ void log_protocol(const char* type,EFI_GUID* guid);
 void* find_protocol(EFI_GUID* guid,EFI_HANDLE handle);
 void install_protocol(EFI_GUID* guid,EFI_HANDLE handle,void* interface);
 intptr_t count_handles(EFI_GUID* guid);
+void* get_variable(EFI_GUID* guid,char16_t* name,unsigned int* data_size,uint32_t* attributes);
+void set_variable(EFI_GUID* guid,const char16_t* name,void* data,unsigned int data_size,uint32_t attributes);
```
Comment (jethrogb) [inline efiperun/main.h]: I think it would be best to use mostly EFI types here, but I don't really care.

**Comment by jethrogb** [comment]
Looks good, thanks!

I will merge after the check is added.

**Comment by ghost** [comment]
I've updated the patch to prefer EFI types and to check `data_size` before writing to it. (I've also updated the sample code above to match.) Let me know what you think.

**Comment by jethrogb** [comment]
Oh for the types I just meant in the function signature. I liked the use of u16string actually. It doesn't matter. This is fine.

**Comment by jethrogb** [comment]
Thanks again!

**Pull request merged**